rbac/management/authorization/token_validator.py

```diff
@@ -0,0 +1,147 @@
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+
+"""A token instrospector class which validates that the given token is valid."""
+import logging
+import re
+
+import requests
+from django.conf import settings
+from joserfc import jwt
+from joserfc.jwk import KeySet
+from joserfc.jwt import JWTClaimsRegistry, Token
+from rest_framework import status
+from rest_framework.request import Request
+
+from .invalid_token import InvalidTokenError
+from .missing_authorization import MissingAuthorizationError
+from .unable_meet_prerequisites import UnableMeetPrerequisitesError
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+# The audience claim we are expecting to find in the token.
+AUDIENCE_CLAIM = "cloud-services"
+# The service accounts claim we are expecting to find in the token.
+SERVICE_ACCOUNTS_CLAIM = "api.iam.service_accounts"
+
+
+class ITSSOTokenValidator:
+    """JWT token  validator."""
+
+    def __init__(self):
+        """Get the OIDC configuration URL."""
+        # TODO replace it with:
+        it_host = settings.IT_SERVICE_HOST
+        it_port = settings.IT_SERVICE_PORT
+        it_scheme = settings.IT_SERVICE_PROTOCOL_SCHEME
+
+        # it_host = "https://sso.stage.redhat.com"
+
+        # The host contains the URL including the port...
+        self.host = f"{it_scheme}://{it_host}:{it_port}/auth/realms/redhat-external"
+        # ... but the issuer does not. We need to make this distinction so that when validating the token, the issuer
+        # correctly matches.
+        self.issuer = f"{it_scheme}://{it_host}/auth/realms/redhat-external"
+        self.oidc_configuration_url = f"{self.host}/.well-known/openid-configuration"
+
+    def validate_token(self, request: Request) -> str:
+        """Validate the JWT token issued by Red Hat's SSO.
+
+        Performs validations on the issuer, audience and scope of the token. Raises exceptions if the token is not
+        valid. Finally, it returns the received Bearer token in order to be able to forward it as is to IT.
+        """
+        if settings.IT_BYPASS_TOKEN_VALIDATION:
+            return "mocked-invalid-bearer-token-because-token-validation-is-disabled"
+
+        bearer_token: str = request.headers.get("Authorization")
+        if not bearer_token:
+            logger.debug(
+                f"Issuing unauthorized response either because the Authorization header is missing, or it's blank."
+                f" Request: {request}"
+            )
+            raise MissingAuthorizationError()
+
+        # Strip the "Bearer" part of the token if it comes with it.
+        if bearer_token.startswith("Bearer"):
+            bearer_token = re.sub("Bearer\\s+", "", bearer_token)
+
+        # Attempt getting IT's OIDC configuration.
+        oidc_response = requests.get(url=self.oidc_configuration_url)
```

Comment:
By default, I think the requests package does not set a timeout on the http calls.  If the call hangs, you might start eating up the threads in the thread pool.

https://requests.readthedocs.io/en/latest/user/quickstart/#timeouts
https://requests.readthedocs.io/en/latest/user/advanced/#timeouts

Reply:
Refactored!